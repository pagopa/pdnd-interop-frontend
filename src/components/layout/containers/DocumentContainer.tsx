import React, { useRef, useState } from 'react'
import type { SxProps } from '@mui/material'
import { InputAdornment, Stack, Tooltip, Button, TextField } from '@mui/material'
import DeleteIcon from '@mui/icons-material/DeleteOutline'
import DownloadIcon from '@mui/icons-material/Download'
import ModeEditIcon from '@mui/icons-material/ModeEdit'
import AttachFileIcon from '@mui/icons-material/AttachFile'
import { useTranslation } from 'react-i18next'
import { InputWrapper } from '@/components/shared/InputWrapper'
import type { EServiceDoc } from '@/api/api.generatedTypes'

type DocumentContainerProps = {
  doc: EServiceDoc
  onDownload?: (document: EServiceDoc) => void
  onDelete?: (document: EServiceDoc) => void
  onUpdateDescription?: (newDescription: string) => void
  sx?: SxProps
}

export function DocumentContainer({
  doc,
  onDownload,
  onDelete,
  onUpdateDescription,
  sx,
}: DocumentContainerProps) {
  const { t } = useTranslation('shared-components', { keyPrefix: 'documentContainer' })
  const inputRef = useRef<HTMLInputElement>(null)
  const [canEdit, setCanEdit] = useState(false)
  const [newValue, setNewValue] = useState(doc.prettyName)

  const updateCanEdit = async (e: React.SyntheticEvent) => {
    e.preventDefault()
    setCanEdit((prev) => {
      const newState = !prev
      if (newState && inputRef && inputRef.current) {
        setTimeout(() => {
          ;(inputRef.current as HTMLInputElement).focus()
        }, 0)
      }
      return newState
    })
  }

  const handleUpdateDescription = () => {
    if (newValue !== doc.prettyName) {
      onUpdateDescription?.(newValue)
    }
    setCanEdit(false)
  }

  return (
    <Stack sx={sx} direction="row" justifyContent="space-between">
      <InputWrapper
        sx={{ my: 0, width: '100%' }}
        infoLabel={canEdit ? t('prettyName.infoLabel') : undefined}
      >
        <TextField
          inputRef={inputRef}
          disabled={!canEdit}
          sx={{ my: 0, width: '100%', flexShrink: 1 }}
          name="prettyName"
          label={t('prettyName.label')}
          value={!canEdit ? doc.prettyName : newValue}
          onChange={(e) => setNewValue(e.target.value)}
          onBlur={handleUpdateDescription}
          inputProps={{ maxLength: 60 }}
          InputProps={{
            startAdornment: (
              <InputAdornment position="start">
                <AttachFileIcon fontSize="small" />
              </InputAdornment>
            ),
          }}
        />
      </InputWrapper>

      <Stack direction="row" alignItems="start" sx={{ flexShrink: 0, ml: 1, mt: 0.8 }}>
        {onUpdateDescription && (
          <Tooltip title={t('editDocumentName')}>
            <Button
              sx={{
                p: 1,
                bgcolor: canEdit ? 'primary.main' : 'transparent',
                color: canEdit ? 'common.white' : 'primary.main',
              }}
              onClick={updateCanEdit}
            >
              <ModeEditIcon fontSize="small" />
            </Button>
          </Tooltip>
        )}
        {onDownload && (
          <Tooltip title={t('downloadDocument')}>
            <Button sx={{ p: 1 }} onClick={onDownload.bind(null, doc)}>
              <DownloadIcon fontSize="small" />
            </Button>
          </Tooltip>
        )}
        {onDelete && (
          <Tooltip title={t('deleteDocument')}>
            <Button color="error" sx={{ p: 1 }} onClick={onDelete.bind(null, doc)}>
              <DeleteIcon fontSize="small" />
            </Button>
          </Tooltip>
        )}
      </Stack>
    </Stack>
  )
}
