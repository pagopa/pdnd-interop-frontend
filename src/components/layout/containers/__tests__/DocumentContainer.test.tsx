import React from 'react'
import { render } from '@testing-library/react'
import { DocumentContainer } from '../DocumentContainer'
import type { EServiceDoc } from '@/api/api.generatedTypes'
import { vi } from 'vitest'
import userEvent from '@testing-library/user-event'

const docMock: EServiceDoc = {
  id: '1',
  name: 'name',
  contentType: 'pdf',
  prettyName: 'document',
}

describe('DocumentContainer', () => {
  it('should match snapshot', () => {
    const screen = render(<DocumentContainer doc={docMock} />)
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should match snapshot with onUpdateDescription callback passed', () => {
    const screen = render(<DocumentContainer doc={docMock} onUpdateDescription={vi.fn()} />)
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should match snapshot with onDelete callback passed', () => {
    const screen = render(<DocumentContainer doc={docMock} onDelete={vi.fn()} />)
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should match snapshot with onDownload callback passed', () => {
    const screen = render(<DocumentContainer doc={docMock} onDownload={vi.fn()} />)
    expect(screen.baseElement).toMatchSnapshot()
  })

  it('should correctly edit document name', async () => {
    const onUpdateDescription = vi.fn()
    const screen = render(
      <DocumentContainer doc={docMock} onUpdateDescription={onUpdateDescription} />
    )
    const user = userEvent.setup()

    const editButton = screen.getByRole('button', { name: 'editDocumentName' })
    await user.click(editButton)
    const input = screen.getByRole('textbox', { name: 'prettyName.label' })
    await user.type(input, ' new name')
    await user.tab()
    expect(onUpdateDescription).toBeCalledWith('document new name')
  })
})
