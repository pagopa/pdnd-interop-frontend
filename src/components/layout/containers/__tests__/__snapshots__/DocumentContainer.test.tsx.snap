// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DocumentContainer > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="css-gmwslw-MuiStack-root"
    >
      <div
        class="MuiBox-root css-n2ek5d"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-8rphz-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r0:"
            id=":r0:-label"
          >
            prettyName.label
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="AttachFileIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.5 6v11.5c0 2.21-1.79 4-4 4s-4-1.79-4-4V5c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5v10.5c0 .55-.45 1-1 1s-1-.45-1-1V6H10v9.5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V5c0-2.21-1.79-4-4-4S7 2.79 7 5v12.5c0 3.04 2.46 5.5 5.5 5.5s5.5-2.46 5.5-5.5V6h-1.5z"
                />
              </svg>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
              disabled=""
              id=":r0:"
              maxlength="60"
              name="prettyName"
              type="text"
              value="document"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  prettyName.label
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="css-qkqpx0-MuiStack-root"
      />
    </div>
  </div>
</body>
`;

exports[`DocumentContainer > should match snapshot with onDelete callback passed 1`] = `
<body>
  <div>
    <div
      class="css-gmwslw-MuiStack-root"
    >
      <div
        class="MuiBox-root css-n2ek5d"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-8rphz-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r3:"
            id=":r3:-label"
          >
            prettyName.label
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="AttachFileIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.5 6v11.5c0 2.21-1.79 4-4 4s-4-1.79-4-4V5c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5v10.5c0 .55-.45 1-1 1s-1-.45-1-1V6H10v9.5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V5c0-2.21-1.79-4-4-4S7 2.79 7 5v12.5c0 3.04 2.46 5.5 5.5 5.5s5.5-2.46 5.5-5.5V6h-1.5z"
                />
              </svg>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
              disabled=""
              id=":r3:"
              maxlength="60"
              name="prettyName"
              type="text"
              value="document"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  prettyName.label
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="css-qkqpx0-MuiStack-root"
      >
        <button
          aria-label="deleteDocument"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeMedium MuiButton-textSizeMedium css-15suzu7-MuiButtonBase-root-MuiButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="DeleteOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`DocumentContainer > should match snapshot with onDownload callback passed 1`] = `
<body>
  <div>
    <div
      class="css-gmwslw-MuiStack-root"
    >
      <div
        class="MuiBox-root css-n2ek5d"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-8rphz-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r5:"
            id=":r5:-label"
          >
            prettyName.label
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="AttachFileIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.5 6v11.5c0 2.21-1.79 4-4 4s-4-1.79-4-4V5c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5v10.5c0 .55-.45 1-1 1s-1-.45-1-1V6H10v9.5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V5c0-2.21-1.79-4-4-4S7 2.79 7 5v12.5c0 3.04 2.46 5.5 5.5 5.5s5.5-2.46 5.5-5.5V6h-1.5z"
                />
              </svg>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
              disabled=""
              id=":r5:"
              maxlength="60"
              name="prettyName"
              type="text"
              value="document"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  prettyName.label
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="css-qkqpx0-MuiStack-root"
      >
        <button
          aria-label="downloadDocument"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1carnhg-MuiButtonBase-root-MuiButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="DownloadIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M5 20h14v-2H5v2zM19 9h-4V3H9v6H5l7 7 7-7z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`DocumentContainer > should match snapshot with onUpdateDescription callback passed 1`] = `
<body>
  <div>
    <div
      class="css-gmwslw-MuiStack-root"
    >
      <div
        class="MuiBox-root css-n2ek5d"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-8rphz-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r1:"
            id=":r1:-label"
          >
            prettyName.label
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="AttachFileIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.5 6v11.5c0 2.21-1.79 4-4 4s-4-1.79-4-4V5c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5v10.5c0 .55-.45 1-1 1s-1-.45-1-1V6H10v9.5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V5c0-2.21-1.79-4-4-4S7 2.79 7 5v12.5c0 3.04 2.46 5.5 5.5 5.5s5.5-2.46 5.5-5.5V6h-1.5z"
                />
              </svg>
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
              disabled=""
              id=":r1:"
              maxlength="60"
              name="prettyName"
              type="text"
              value="document"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  prettyName.label
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="css-qkqpx0-MuiStack-root"
      >
        <button
          aria-label="editDocumentName"
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-n9wclb-MuiButtonBase-root-MuiButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="ModeEditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
